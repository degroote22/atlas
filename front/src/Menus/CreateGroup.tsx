import * as React from "react";
import GroupStore from "../Stores/GroupStore";

interface IProps {
  GroupStore?: typeof GroupStore;
}

const initialState = {
  newTitle: "",
  error: false
};

type IState = Readonly<typeof initialState>;

class CreateGroup extends React.Component<IProps, IState> {
  public static defaultProps: Partial<IProps> = {
    GroupStore: GroupStore
  };

  readonly state: IState = initialState;

  private GroupStore() {
    if (!this.props.GroupStore) {
      throw Error(
        "Group Store não definida em Create Group"
      );
    }
    return this.props.GroupStore;
  }

  private onChangeTitle = (event: any) => {
    this.setState({ newTitle: event.target.value });
  };

  private onCreate = () => {
    if (this.isDisabled()) {
      return;
    }
    this.setState({
      error: false
    });
    this.GroupStore()
      .create(this.state.newTitle)
      .then(() => {
        this.setState(initialState);
      })
      .catch(() => {
        this.setState({ error: true });
      });
  };

  private renderError() {
    return (
      this.state.error && (
        <div className="_error field has-addons">
          <p className="subtitle">
            Houve um erro no upload!
          </p>
        </div>
      )
    );
  }

  private isDisabled = () =>
    this.state.newTitle.length === 0;

  render() {
    return (
      <section className="hero is-info">
        <div className="hero-body">
          <div className="container">
            <h1 className="title">Criar nova seção</h1>
            <div className="field has-addons">
              <div className="control">
                <input
                  onChange={this.onChangeTitle}
                  value={this.state.newTitle}
                  className="input _input"
                  type="text"
                  placeholder="Digite o nome"
                />
              </div>
              <div className="control">
                <button
                  disabled={this.isDisabled()}
                  className="button is-dark _create"
                  onClick={this.onCreate}
                >
                  Criar
                </button>
              </div>
            </div>
            {this.renderError()}
          </div>
        </div>
      </section>
    );
  }
}

export default CreateGroup;
