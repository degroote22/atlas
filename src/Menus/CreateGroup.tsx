import * as React from "react";
interface IProps {
  loading: boolean;
  onCreateGroup: (title: string) => void;
}

interface IState {
  newTitle: string;
}

class CreateGroup extends React.Component<IProps, IState> {
  state = {
    newTitle: ""
  };

  private onChangeTitle = (event: any) => {
    this.setState({ newTitle: event.target.value });
  };

  private onCreate = () => {
    this.props.onCreateGroup(this.state.newTitle);
  };

  render() {
    return (
      <section className="hero is-success">
        <div className="hero-body">
          <div className="container">
            <h1 className="title">Criar nova seção</h1>
            <div className="field has-addons">
              <div className="control">
                <input
                  onChange={this.onChangeTitle}
                  value={this.state.newTitle}
                  className="input"
                  type="text"
                  placeholder="Digite o nome"
                />
              </div>
              <div className="control">
                <a
                  className="button is-dark"
                  onClick={this.onCreate}
                >
                  Criar
                </a>
              </div>
            </div>
          </div>
        </div>
      </section>
    );
  }
}

export default CreateGroup;
