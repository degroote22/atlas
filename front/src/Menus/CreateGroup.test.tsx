import * as React from "react";
import * as ReactDOM from "react-dom";
import CreateGroup from "./CreateGroup";
import * as renderer from "react-test-renderer";
import { shallow } from "enzyme";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<CreateGroup />, div);
});

it("renders correctly", () => {
  const tree = renderer.create(<CreateGroup />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("does not create a group without a title", () => {
  const mock = jest
    .fn()
    .mockReturnValueOnce(Promise.resolve());

  const wrapper = shallow(
    <CreateGroup
      GroupStore={
        {
          create: mock
        } as any
      }
    />
  );

  const title = "";

  wrapper
    .find("._input")
    .simulate("change", { target: { value: title } });

  wrapper.find("._create").simulate("click");

  expect(mock).not.toHaveBeenCalled();
});

it("creates a group and clears state", done => {
  const mock = jest
    .fn()
    .mockReturnValueOnce(Promise.resolve());
  const wrapper = shallow(
    <CreateGroup
      GroupStore={
        {
          create: mock
        } as any
      }
    />
  );

  const title = "Novo TÃ­tulo";

  wrapper
    .find("._input")
    .simulate("change", { target: { value: title } });

  wrapper.find("._create").simulate("click");
  wrapper.update();
  expect(wrapper.instance().state.newTitle).toBe(title);
  expect(wrapper.getElement()).toMatchSnapshot();
  expect(mock).toHaveBeenCalledWith(title);

  setTimeout(() => {
    wrapper.update();
    expect(wrapper.instance().state.newTitle).toBe("");
    expect(wrapper.instance().state.error).toBe(false);
    expect(wrapper.getElement()).toMatchSnapshot();
    done();
  }, 0);
});

it("displays an error if needed", done => {
  const mock = jest
    .fn()
    .mockReturnValueOnce(Promise.reject("_ANY_"));

  const wrapper = shallow(
    <CreateGroup
      GroupStore={
        {
          create: mock
        } as any
      }
    />
  );

  const title = "XXXXXXXXXXXXXxx";

  wrapper
    .find("._input")
    .simulate("change", { target: { value: title } });

  wrapper.find("._create").simulate("click");

  expect(mock).toHaveBeenCalledWith(title);

  setTimeout(() => {
    wrapper.update();
    expect(wrapper.instance().state.error).toBe(true);
    expect(wrapper.find("._error")).toHaveLength(1);
    expect(
      wrapper.find("._error").getElement()
    ).toMatchSnapshot();
    done();
  }, 0);
});
