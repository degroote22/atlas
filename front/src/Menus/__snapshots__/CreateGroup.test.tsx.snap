// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a group and clears state 1`] = `
<section
  className="hero is-info"
>
  <div
    className="hero-body"
  >
    <div
      className="container"
    >
      <h1
        className="title"
      >
        Criar nova seção
      </h1>
      <div
        className="field has-addons"
      >
        <div
          className="control"
        >
          <input
            className="input _input"
            onChange={[Function]}
            placeholder="Digite o nome"
            type="text"
            value="Novo Título"
          />
        </div>
        <div
          className="control"
        >
          <button
            className="button is-dark _create"
            disabled={false}
            onClick={[Function]}
          >
            Criar
          </button>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`creates a group and clears state 2`] = `
<section
  className="hero is-info"
>
  <div
    className="hero-body"
  >
    <div
      className="container"
    >
      <h1
        className="title"
      >
        Criar nova seção
      </h1>
      <div
        className="field has-addons"
      >
        <div
          className="control"
        >
          <input
            className="input _input"
            onChange={[Function]}
            placeholder="Digite o nome"
            type="text"
            value=""
          />
        </div>
        <div
          className="control"
        >
          <button
            className="button is-dark _create"
            disabled={true}
            onClick={[Function]}
          >
            Criar
          </button>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`displays an error if needed 1`] = `
<div
  className="_error field has-addons"
>
  <p
    className="subtitle"
  >
    Houve um erro no upload!
  </p>
</div>
`;

exports[`renders correctly 1`] = `
<section
  className="hero is-info"
>
  <div
    className="hero-body"
  >
    <div
      className="container"
    >
      <h1
        className="title"
      >
        Criar nova seção
      </h1>
      <div
        className="field has-addons"
      >
        <div
          className="control"
        >
          <input
            className="input _input"
            onChange={[Function]}
            placeholder="Digite o nome"
            type="text"
            value=""
          />
        </div>
        <div
          className="control"
        >
          <button
            className="button is-dark _create"
            disabled={true}
            onClick={[Function]}
          >
            Criar
          </button>
        </div>
      </div>
    </div>
  </div>
</section>
`;
